### PR TITLE
State Management Port from JS Code

### PR DESCRIPTION
**Pull request by cleemullins**

### CODE REVIEW

File: tests/Microsoft.Bot.Builder.Tests/SlotTests.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        public void Ctor_Should_Register_SlotDefinition()
+        {
+            var mockFrame = new Mock<IFrame>();                        
+            var sd = new SlotDefinition()
+            {
+                Name = "testName",
+            };
+
+            var x = new Slot(mockFrame.Object, sd);
+            Assert.IsTrue(x.Definition.Name == "testName");
```
Comment (drub0y) [inline tests/Microsoft.Bot.Builder.Tests/SlotTests.cs]: Should prefer the more specific `Assert.AreEqual` for equality checks.

File: tests/Microsoft.Bot.Builder.Tests/SlotTests.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        }
+
+        [TestMethod]
+        public void Ctor_Should_Register_The_Frame()
+        {
+            var mockFrame = new Mock<IFrame>();
+            var sd = new SlotDefinition();
+            var x = new Slot(mockFrame.Object, sd);                 
+            mockFrame.Verify(t => t.AddSlot(x));
+        }
```
Comment (drub0y) [inline tests/Microsoft.Bot.Builder.Tests/SlotTests.cs]: Should probably assert that the `Slot.Frame` property is equal to the `mockFrame.Object` as well? Probably more important than checking that `AddSlot` was called which is kind of an internal implementation detail.

File: tests/Microsoft.Bot.Builder.Tests/SlotTests.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+using System;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Moq;
+
+namespace Microsoft.Bot.Builder.Tests
+{
+    [TestClass]
+    [TestCategory("State")]
+    public class SlotTests
+    {
```
Comment (drub0y) [inline tests/Microsoft.Bot.Builder.Tests/SlotTests.cs]: Instead of all of these being prefixed with `Ctor_` I would recommend following the pattern of introducing a nested class for the method under test... in this case `public class ConstructorTests` and then putting the test methods inside of there. This works nicely with test runners such as VS's (finally) that understand the hierarchial test class pattern. This also means you move the `TestClass`/`TestCategory` attributes from the `SlotTests` parent class and put it on the nested classes instead... this is ceremony overhead of MSTest, XUnit you wouldn't have to apply an attribute the classes at all.

NOTE: this suggestion applies to all the tests in this PR, but I'm only writing this one comment.

File: tests/Microsoft.Bot.Builder.Tests/FrameDefinitionTests.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+            {
+                Scope = targetScope,
+                NameSpace = targetNs,
+            };
+
+            Assert.AreEqual(targetScope, fd.Scope);
+            Assert.AreEqual(targetNs, fd.NameSpace);
+
+            // Should defult to empty. 
+            Assert.IsTrue(fd.SlotDefinitions.Count == 0);
```
Comment (drub0y) [inline tests/Microsoft.Bot.Builder.Tests/FrameDefinitionTests.cs]: Should prefer the more specific Assert.AreEqual for equality checks.

File: tests/Microsoft.Bot.Builder.Tests/RootFrameTests.cs [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        public void Ctor_Should_Throw_On_NullStorage()
+        {
+            var x = new RootFrame(null, new FrameDefinition());
+        }
+
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentNullException))]
+        public void Ctor_Should_Throw_On_NullFrameDefinition()
+        {            
+            var x = new RootFrame(new Mock<IStorage>().Object, null);
```
Comment (drub0y) [inline tests/Microsoft.Bot.Builder.Tests/RootFrameTests.cs]: This pattern can be simplified. Instead of having to do `new Mock<Whatever>.Object` you can just write `Mock.Of<Whatever>()`.

NOTE: only putting this one comment on, but applies everywhere the pattern is being used.

File: tests/Microsoft.Bot.Builder.Tests/RootFrameTests.cs [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+using System.Collections.Generic;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Moq;
+
+namespace Microsoft.Bot.Builder.Tests
+{
+    [TestClass]
+    [TestCategory("State")]
+    public class RootFrameTests
+    {
```
Comment (drub0y) [inline tests/Microsoft.Bot.Builder.Tests/RootFrameTests.cs]: See other comment on hierarchical test pattern. Every set of tests for a particular method here should get a nested class to avoid the `MethodName_` prefixing... plays better with modern test runners.

File: libraries/Microsoft.Bot.Builder/IReadWriteSlot.cs [hunk 7]
```diff
@@ -7,7 +7,7 @@ namespace Microsoft.Bot.Builder
 {
     public interface IReadWriteSlot : IReadOnlySlot
```
Comment (drub0y) [inline libraries/Microsoft.Bot.Builder/IReadWriteSlot.cs]: Needs doc comments.

File: libraries/Microsoft.Bot.Builder/CachedFrameState.cs [hunk 8]
```diff
@@ -0,0 +1,14 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace Microsoft.Bot.Builder
+{
+    public class CachedFrameState
```
Comment (drub0y) [inline libraries/Microsoft.Bot.Builder/CachedFrameState.cs]: Needs doc comments.

File: libraries/Microsoft.Bot.Builder/FrameDefinition.cs [hunk 9]
```diff
@@ -0,0 +1,13 @@
+﻿using System.Collections.Generic;
+
+namespace Microsoft.Bot.Builder
+{
+    public class FrameDefinition
```
Comment (drub0y) [inline libraries/Microsoft.Bot.Builder/FrameDefinition.cs]: Needs doc comments.

File: libraries/Microsoft.Bot.Builder/FrameDefinition.cs [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+    {
+        public string Scope { get; set; }
+
+        public string NameSpace { get; set; }
+
+        public List<SlotDefinition> SlotDefinitions { get; set; } = new List<SlotDefinition>();
```
Comment (drub0y) [inline libraries/Microsoft.Bot.Builder/FrameDefinition.cs]: I would not personally allow a `List<T>` property to be `set`-able instead preferring to expose as read-only and prefer callers to just use the `List<T>` API to mutate the contents.

File: libraries/Microsoft.Bot.Builder/IFrame.cs [hunk 11]
```diff
@@ -1,19 +1,18 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
-using System.Collections.Generic;
 using System.Threading.Tasks;
 
 namespace Microsoft.Bot.Builder
 {
     public interface IFrame
```
Comment (drub0y) [inline libraries/Microsoft.Bot.Builder/IFrame.cs]: Needs doc comments.

File: libraries/Microsoft.Bot.Builder/IFrame.cs [hunk 12]
```diff
@@ -1,19 +1,18 @@
...
     {
-        IFrame Parent { get; set; }
+        IFrame Parent { get; }
 
         void AddSlot(IReadWriteSlot slot);
```
Comment (drub0y) [inline libraries/Microsoft.Bot.Builder/IFrame.cs]: I would highly recommend this design be changed. The coupling between this method and the `Slot` constructor is extremely confusing.

I would instead suggest `Slot` actually be removed entirely as a `public` class and this method be changed to a factory method like so: 

```CSharp
IReadWriteSlot<T> AddSlot<T>(SlotDefinition slotDefinition)
```
Internally the implementation then has the responsibility to construct the now `internal` `Slot` class and introduce itself as the `Frame` and users do not have to deal with that coupling at all.

File: libraries/Microsoft.Bot.Builder/IReadWriteSlot.cs [hunk 13]
```diff
@@ -7,7 +7,7 @@ namespace Microsoft.Bot.Builder
...
     {
```
Comment (drub0y) [inline libraries/Microsoft.Bot.Builder/IReadWriteSlot.cs]: I believe this interface should just be renamed to simply `ISlot`. An `ISlot` is implicitly fully read/write. Also see comments on `IFrame::AddSlot` as this is what would now be returned from the factory method and people wouldn't actually ever be constructing explicit `Slot` class instances any more.

File: libraries/Microsoft.Bot.Builder/RootFrame.cs [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Microsoft.Bot.Builder
+{
+    public class RootFrame : IFrame
+    {
+        private readonly string _cacheKey = $"RootFrame-{Guid.NewGuid()}";
```
Comment (drub0y) [inline libraries/Microsoft.Bot.Builder/RootFrame.cs]: I think this is a good default, but I believe a constructor overload that allows people to supply their own name for this would be useful for logging/debugging scenarios.

File: libraries/Microsoft.Bot.Builder/Slot.cs [hunk 15]
```diff
@@ -10,15 +10,15 @@ namespace Microsoft.Bot.Builder
 {
     public class Slot : IReadWriteSlot
     {
-        public Slot(IFrame frame, ISlotDefinition definition)
+        public Slot(IFrame frame, SlotDefinition definition)
```
Comment (drub0y) [inline libraries/Microsoft.Bot.Builder/Slot.cs]: This coupling with the `IFrame` will be super confusing for customers. See comments on `IFrame::AddSlot` for suggestions on how to change the API which would include making this entire class an internal implementation detail as they would only ever work with the `IReadWriteSlot` interface.

File: libraries/Microsoft.Bot.Builder/SlotDefinition.cs [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+{
+    public class SlotDefinition
+    {
+        public string Name { get; set; }
+
+        public object DefaultValue { get; set; }
+
+        public string[] ChangeTags { get; set; }
+
+        public int ExpiresAfterSeconds { get; set; }
```
Comment (drub0y) [inline libraries/Microsoft.Bot.Builder/SlotDefinition.cs]: Why is this here if this is on `SlotHistoryPolicy`? Seems like it should only be in one place.

File: libraries/Microsoft.Bot.Builder/SlotDefinition.cs [hunk 17]
```diff
@@ -0,0 +1,18 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace Microsoft.Bot.Builder
+{
+    public class SlotDefinition
```
Comment (drub0y) [inline libraries/Microsoft.Bot.Builder/SlotDefinition.cs]: Needs doc comments.

File: libraries/Microsoft.Bot.Builder/SlotHistoryPolicy.cs [hunk 18]
```diff
@@ -5,8 +5,8 @@ namespace Microsoft.Bot.Builder
 {
     public class SlotHistoryPolicy
     {
-        public int MaxCount { get; set; }
```
Comment (drub0y) [inline libraries/Microsoft.Bot.Builder/SlotHistoryPolicy.cs]: Needs doc comments.

**Comment by cleemullins** [comment]
Due to changes in design, this code is being retired. I'll leave the branch for now (in case this comes back) but the new code will be different enough that starting over is probably the easiest choice.
